File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())
tree = [[] for _ in range(n+1)]

for _ in range(n-1):
    a, b = map(int, input().split())
    tree[a].append(b)
    tree[b].append(a)

visited = [False] * (n + 1)
answer = [0] * (n + 1)
def visit_tree(start_node):
    queue = deque([start_node])
    visited[start_node] = True

    while queue:
        cur_node = queue.popleft()
        for next_node in tree[cur_node]:
            if not visited[next_node]:
                queue.append(next_node)
                visited[next_node] = True
                answer[next_node] = cur_node

visit_tree(1)

for i in range(2, n+1):
    print(answer[i])

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt"
  )
  .toString()
  .split("\n");

const n = Number(input.shift());
const graph = Array.from({ length: n + 1 }, () => []);

for (let i = 0; i < n - 1; i++) {
  const [a, b] = input[i].split(" ").map(Number);
  graph[a].push(b);
  graph[b].push(a);
}

const parent = Array.from({ length: n + 1 }, () => 0);

function bfs(start) {
  const visited = Array.from({ length: n + 1 }, () => false);
  const queue = [];

  visited[start] = true;
  queue.push(start);

  while (queue.length) {
    const cur_node = queue.shift();

    for (let next_node of graph[cur_node]) {
      if (!visited[next_node]) {
        visited[next_node] = true;
        queue.push(next_node);
        parent[next_node] = cur_node;
      }
    }
  }
}

bfs(1);
console.log(parent.slice(2).join("\n"));
